File: 2025/golang/day6/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func performOperation(numbers []int, operation string) int {
	result := 0
	if operation == "+" {
		for _, num := range numbers {
			result += num
		}
	} else if operation == "*" {
		result = 1
		for _, num := range numbers {
			result *= num
		}
	}
	return result
}

func populateArr(arr [][]int, firstRow []string) [][]int {
	for i := 0; i < len(firstRow); i++ {
		if len(strings.ReplaceAll(firstRow[i], " ", "")) > 0 {
			arr = append(arr, []int{})
		}
	}

	return arr
}

func solveDay6Part1(input string) {
	file, err := os.Open(input)
	if err != nil {
		panic(err)
	}

	total := 0
	scanner := bufio.NewScanner(file)
	arr := [][]int{}
	isLastLine := false
	isPopulated := false
	ops := []string{}

	// Get data
	for scanner.Scan() {
		line := scanner.Text()

		numbers := strings.Split(line, " ")

		if !isPopulated {
			arr = populateArr(arr, numbers)
			isPopulated = true
		}

		if numbers[0] == "*" || numbers[0] == "+" {
			isLastLine = true
		}

		j := 0
		for i := 0; i < len(numbers); i++ {
			numTrimmed := strings.ReplaceAll(numbers[i], " ", "")
			if len(numTrimmed) == 0 {
				continue
			}
			if !isLastLine {
				n, err := strconv.Atoi(numTrimmed)
				if err != nil {
					panic(err)
				}
				arr[j] = append(arr[j], n)
			} else {
				if numTrimmed == "+" || numTrimmed == "*" {
					ops = append(ops, numTrimmed)
				}
			}

			j++
		}
	}

	for i, num := range arr {
		total += performOperation(num, ops[i])
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	fmt.Println("Total: ", total)
}

func main() {
	// input := "input_test.txt" // Expected output: 4277556
	input := "input.txt" // 7326876294741
	solveDay6Part1(input)
}

File: 2025/cpp/day6/part1.cpp
#include <iostream>
#include <fstream>
#include <sstream>

long long perform_operation(std::vector<long long>& numbers, std::string operation) {
    long long result = 0;
    if (operation == "+") {
        for (auto num : numbers) {
            result += num;
        }
    } else if (operation == "*") {
        result = 1;
        for (auto num : numbers) {
            result *= num;
        }
    }
    return result;
}

void populate_array(std::vector<std::vector<long long>>& array, std::vector<std::string> first_row) {
    using namespace std;
    for (size_t i = 0; i < first_row.size(); ++i) {
        first_row[i].erase(
            std::remove(first_row[i].begin(), first_row[i].end(), ' '),
            first_row[i].end()
        );
        if (first_row[i].empty()) {
            continue;
        }
        array.push_back(std::vector<long long>());
    }
}

std::vector<std::string> split(const std::string& str, char delimiter) {
    using namespace std;

    vector<string> tokens;
    string token;
    istringstream tokenStream(str);
    while (getline(tokenStream, token, delimiter)) {
        tokens.push_back(token);
    }
    return tokens;
}

void solve_day6_part1(std::string input_file) {
    using namespace std;

    long long total = 0;
    ifstream input(input_file);
    string line;
    vector<vector<long long>> arr;
    int is_last_line = 0;
    int is_populated = 0;
    vector<string> ops;

    while (getline(input, line)) {

        vector<string> numbers = split(line, ' ');

        if (!is_populated) {
            populate_array(arr, numbers);
            is_populated = 1;
        }

        if (numbers[0] == "*" || numbers[0] == "+") {
            is_last_line = 1;
        }

        int j = 0;
        for (size_t i = 0; i < numbers.size(); ++i) {
            string num_trimmed = numbers[i];
            num_trimmed.erase(
                remove(num_trimmed.begin(), num_trimmed.end(), ' '),
                num_trimmed.end()
            );
            if (num_trimmed.empty()) {
                continue;
            }

            if (!is_last_line) {
                arr[j].push_back(stoi(num_trimmed));
            } else {
                if (num_trimmed == "*" || num_trimmed == "+") {
                    ops.push_back(num_trimmed);
                }
            }

            j++;
        }
    }

    for (size_t i = 0; i < arr.size(); ++i) {
        total += perform_operation(arr[i], ops[i]);
    }

    cout << "Total: " << total << endl;
}


int main() {
    // std::string input_file = "input_test.txt"; // Expected output: 4277556
    std::string input_file = "input.txt"; // 7326876294741 
    solve_day6_part1(input_file);
}
